coderd/database/queries/workspaces.sql

```diff
@@ -157,10 +157,12 @@ WHERE
 					latest_build.canceled_at IS NULL AND
 					latest_build.completed_at IS NOT NULL AND
 					latest_build.updated_at - INTERVAL '30 seconds' < NOW() AND
-					latest_build.transition = 'delete'::workspace_transition
+					latest_build.transition = 'delete'::workspace_transition AND
+					-- If the error field is null, the status is 'failed'
+					latest_build.error IS NULL
 
 				WHEN @status = 'deleting' THEN
-					latest_build.completed_at IS NOT NULL AND
+					latest_build.completed_at IS NULL AND
```

Comment:
so, it's really difficult to read this query and convince yourself that these WHERE clauses are 

a) non-overlapping
b) match the golang implementation of the status computation

What do you think about the strategy of introducing a stored procedure that computes the status?  Then the filter logic in this query becomes a trivial WHERE clause, and we can use the DB-computed status in prod, rather than computing the status in the API layer.

Annoyingly, we'd still have to keep the golang implementation around for the dbfake, but they could both be written in an imperative style and so it will be much easier to verify the algorithm matches.

Reply:
Oh it was a headache. Are we using stored procedures anywhere else, or would this be the first case? I'm open to anything that makes this easier.

Ideally though, it would be in the next PR so this bug is fixed in main. The major contribution from this PR is the test imo.

Reply:
We have triggers for deletion of tokens stored in the DB.  This would be the first computed column, though.

Fine to follow up in another PR.